### ISSUE TITLE
Approve every new version of public app

### ISSUE THREAD

**Issue by JakubMatejka** [keboola/developer-portal-ui#491]
Jakmile je komponenta publikovaná, tak se její další úpravy neprojeví navenek dokud vendor znovu nepožádá o publikaci. Na straně API to funguje stejně jako předtím, publishing request se může odesílat opakovaně i po první publikaci (https://kebooladeveloperportal.docs.apiary.io/#reference/0/app/request-publishing)

**Comment by ujovlado** [comment]
@jakubmatejka a zmenili sa nejake responses, ked to treba spravit synchronizovane?

Idealne by bolo, keby som to robil voci realnemu api, ktore je 1:1 s produkciou (testing?). Data samozrejme sediet nemusia.

**Comment by JakubMatejka** [comment]
Nezměnily, ale nepůjde schvalovat appky do KBC (tzn. nemusí to být úplně na minutu, ale ideálně ne moc dlouho po sobě). 

Hm, asi bych k tomu mohl nachystat nějaký "staging" stack? Máme testovací pro Travis, ale to by se zas nemohly integrovat další úpravy než to dokončíte

**Comment by JakubMatejka** [comment]
ui by se tím pádem dalo i nasadit dřív, ale nepůjde v něm vyzkoušet ta další publikace

**Comment by ujovlado** [comment]
Ok, som z toho trosku jelen, tak skusim popisat ako to chapem.

> Jakékoli úpravy komponenty se neprojeví v KBC dokud vendor znovu nepožádá o publikaci. Na straně API to funguje stejně jako předtím, publishing request se může odesílat opakovaně i po první publikaci (https://kebooladeveloperportal.docs.apiary.io/#reference/0/app/request-publishing).

To znamena, ze so ziadostami o publikovanie nemusim robit nic. Akurat teda zmenim notifikaciu pre usera, pretoze obcase to posle na schvalenie a obcas to rovno publikuje.

> Pokud je uživatel admin n. trusted vendor n. je komponenta isPublic=0, tak se ten request ihned publishne a nečeká na náš approve.

Uz som vlastne popisal vysssie

> O tom jestli je komponenta (resp. její verze) publikovaná rozhoduje flag isPublished. V důsledku toho nastavuje isPublic sám uživatel.

Takze do properties apky (do editacie) potrebujem dat flag `isPublic` aby ho user mohol nastavovat.

**Comment by JakubMatejka** [comment]
jj. ale počkal bych ještě na @jancermak (viz. Slack)

**Comment by jancermak** [comment]
@JakubMatejka  — dotaz.
![kdp-prototyp 2018-06-20 16-55-15](https://user-images.githubusercontent.com/1826442/41666641-3f91c29e-74ab-11e8-9eec-df9e5ed2dcbc.png)

**Comment by JakubMatejka** [comment]
Dvojí akce určitě ne, všechny se dělají na appce samotné, ne jejích verzích.
A ty detaily právě nevím jestli mít v seznamu odkazy zvlášť na jednu či druhou (tu publishnutou teda jen read-only), nebo jestli by to šlo nějak rozumně dát vedle sebe do jednoho detailu?

**Comment by ujovlado** [comment]
Osobne by som to navrhoval takto:

- editacia vzdy pojde na poslednu verziu komponenty (podobne je na tom publikovanie, t.j. ked requestnem publikovanie komponenty, mal by som chciet publikovat poslednu verziu, nie nieco historicke)
- pridame moznost pozriet si diff medzi poslednou verziou komponenty a publikovanou

**Comment by ujovlado** [comment]
K tomu co si navrhol vyssie:

- versions last/public -> super, pridal by som tam este niekde tlacitko na diff (komponentu na diff, uz mame, mozem ti to sem pastnut)
- Edit/Remove/Deprecate -> tak ako pise Jakub, tie akcie su nad komponentou, nie nad jej verziou.

Zaroven by som nemal vidiet "Request publishing" ak su verzie rovnake.

@JakubMatejka publikovanie komponenty spravi novu verziu?

**Comment by jancermak** [comment]
@ujovlado @JakubMatejka 
1)components upraveno
2) navrh trusted vendora + jsem to tam trochu posoupal - bo mi ted vychazi ten link na eboola.com... :-)
3)diff — po kliknuti na nazev komponenty se dostanu na jeji detail co tam k tlacitku edit pridat tlacitko edit and compare - ktere mi zapne editace te nove verze a vedle toho zobrazi public verzi s vyznacenymi rozdily?

![image](https://user-images.githubusercontent.com/1826442/41670587-00dfaef8-74b5-11e8-9104-40f06dda236c.png)

![185 - uprava logiky components](https://user-images.githubusercontent.com/1826442/41670662-2e50158a-74b5-11e8-9681-5bcaa9f45ff4.png)

![185 - uprava logiky trusted vendors](https://user-images.githubusercontent.com/1826442/41670671-34e86b22-74b5-11e8-98ad-be91ca3324b8.png)

**Comment by ujovlado** [comment]
@jancermak ten "trusted" flag som na prvy pokus hladal v statuse, myslis, ze by sa to nehodilo tam? (myslim pod "Public, Approved")

**Comment by ujovlado** [comment]
> 2) navrh trusted vendora + jsem to tam trochu posoupal - bo mi ted vychazi ten link na eboola.com... :-)

toto ma sere, fixnem to cim skor, bude tam len nieco ako "EU - 1234" alebo "US - 4456"

**Comment by ujovlado** [comment]
> 3)diff — po kliknuti na nazev komponenty se dostanu na jeji detail co tam k tlacitku edit pridat tlacitko edit and compare - ktere mi zapne editace te nove verze a vedle toho zobrazi public verzi s vyznacenymi rozdily?

diff by sa mi hodil na zozname komponent. pretoze odtial ziadam aj publikovanie ... t.j. mozno spravit z cisla posledne verzie link? alebo do dat so zatvorky za cislo poslednej verzie, ked sa lisia?

**Comment by jancermak** [comment]
@ujovlado 
![image](https://user-images.githubusercontent.com/1826442/41671267-e2cc0f90-74b6-11e8-856e-6ec1aaf87ba2.png)

**Comment by ujovlado** [comment]
parada :+1:

**Comment by ujovlado** [comment]
btw, ten diff mame v admin casti ako modal, vyzera to takto:

![screenshot_2018-06-20_18-23-27](https://user-images.githubusercontent.com/419849/41671358-2d92b2fe-74b7-11e8-912c-34dafd185194.png)

t.j. moj navrh je zo zoznamu apiek na "nieco" kliknut a potom by som videl takyto modal, ked si pozriem zmeny. ak som s nimi spokojny, modal zavriem a stlacim request publishing.

**Comment by jancermak** [comment]
@ujovlado — aha, takhle to funguje. Co sedy okynko to nejaka zmena, nadtim je napsan title. A to cerveny znamena co sem smazal a zeleny co to nahradilo.

**Comment by ujovlado** [comment]
@jancermak presne. s tym, ze by hore nebolo "Changes" ale "Comparing latest (34) and published (22) version". Samozrejmem nemusi to byt nutne takto - len popisujem ako by som to robil, keby som nemal design/navrh od teba.

Ak mas moznost prepnut sa ako admin, tak v zozname komponent je tlacitko "Changes", kde sa dostanes na zmeny za poslednych 28 dni. No a odtial sa kliika na diff. A my by sme "vykradli" len ten diff ;)

**Comment by jancermak** [comment]
@ujovlado ok — ja mel v hlave klasicky komperovani ze dam vedle sebe ty dve verze a ukazu rozdily. Ale tuto mi prijde cajk.

Posilam nahled, kde by mohlo byt to tlacitko. (Co sem pogooglil tato ikonka se pry v GITU pouziva pro compare)
![image](https://user-images.githubusercontent.com/1826442/41672682-d5407ed4-74ba-11e8-9636-d0795860b782.png)

**Comment by JakubMatejka** [comment]
No, já chtěl argumentovat, že diff je příliš nerdy, ale tak je to pro developery, tak proč ne, když se to nějak hezky nakreslí.
Já si akorát nejsem moc jistej tím zobrazením publikování ve výpisu, to už je návrh, nebo to budeš ještě řešit? Já bych ten status private/public od publikování víc oddělil, to už spolu nebude vůbec souviset, viz ten můj popisek ve slacku

**Comment by Halama** [comment]
me se to zatim cely takhle nezda, vypada to tam hrozne dobastleny. Zitra bych zkusil rozepsat nebo se o tom muzeme pobavit.

**Comment by jancermak** [comment]
@JakubMatejka
Tuto je ode me rozpracovany, ujasnuju si zadani. 

Ohledne  toho private/ public jsi jeste na prijmu na rychlo call?

**Comment by ujovlado** [comment]
@JakubMatejka aha jo, chapem ... @jancermak to mozno z toho uplne nepochopil. ;)

Private/Public je teda status komponenty (ci bude pre vsetkych alebo skryta)

V tom prvom riadku ako su verzie 23/42, tak:
- 42 je posledna verzia
- 23 je publikovana verzia

^ a nema to vplyv na to, ci je komponenta private/public. Snad je to tak zrozumitelnejsie

**Comment by jancermak** [comment]
@JakubMatejka @ujovlado @Halama 

S jakubem po telefonu domluveno:

1) private / public se prepina v detailu komponenty
2) last verze — ma u sebe "tlacitko" na jeji odeslani ke schvaleni —> u toho resime textaci - ted tam je "request publishing", napadlo me "push changes"
3) nakreslim jeste detail komponenty s viditelnym comparovanim jako alternativu k ujovu navrhu modalu.
![image](https://user-images.githubusercontent.com/1826442/41674135-667f793c-74bf-11e8-80dc-69d533d4c705.png)

**Comment by ujovlado** [comment]
Imho tam este stale "zavádza" to, ze status (private/public) je v tom skombinovany.

Ak si to prehodis a status bude v samostatnom stlpci, bude to pochopitelnejsie - skusim sem pastnut.

**Comment by jancermak** [comment]
@ujovlado 
![image](https://user-images.githubusercontent.com/1826442/41674805-46a97af2-74c1-11e8-9247-5abff1d0fd3e.png)

**Comment by ujovlado** [comment]
Jop, akurat som nahodil presne nieco podobne ;)

![screenshot_2018-06-20_19-39-09](https://user-images.githubusercontent.com/419849/41674974-aa95ff9a-74c1-11e8-8ab2-66ee10be7314.png)

**Comment by ujovlado** [comment]
^ je to len pre ilutraciu. a chyba mi tam tiez to, ze ked su verzie rovnake, nie je nutne ich tam zobrazovat dvojmo (nemas s cim spravit diff).

\+ ked vieme rozoznat ci sa jedna o trusted vendor alebo o apku ktora je private vieme namiesto "Req. publishing" dat rovno "Publish", lebo to sa naozaj stane.

**Comment by jancermak** [comment]
@ujovlado 
![image](https://user-images.githubusercontent.com/1826442/41675486-19cf59c8-74c3-11e8-919d-a06e4cef975a.png)

**Comment by ujovlado** [comment]
^ S tymto screenom som celkom spoko. Uvidime este, co chcel doriesit Martin. :)

**Comment by ujovlado** [comment]
Potom este doladime texty. V hlavicke by malo pri verziach byt myslim "Last/Published" a tiez by som navrhoval asi premenovat "Status" - lebo je to skor o tom, ci je komponenta pre vsetkych (public) alebo skryta (private).

Sakra a este som zabudol jednu vec - komponente moze niekto z nas zamietnut publikovanie. Vtedy sa tam vyplna ten rejection reason (vyplna to admin). A userovi sa nasledne v komponente zobrazi "Rejected" (a to male "i"). To by malo byt potom pri aktualnej verzii namiesto "Req. publishing".

![screenshot_2018-06-20_20-37-35](https://user-images.githubusercontent.com/419849/41677839-c8a49ebc-74c9-11e8-8d21-62d454983a0b.png)

**Comment by jancermak** [comment]
@ujovlado — u toho rejected ma byt jeste "Request Publishing" ?? tak jak je to ted? Nebo jinak jak udelam z rejectovanyho stav request publishing? — Udelam to tim ze neco upravim a tim mi vznikne nova verze? (tedy to request publishing u rejectu neni potreba?)

![image](https://user-images.githubusercontent.com/1826442/41967580-32cee436-7a02-11e8-8ccf-53daf5d0ad5f.png)

![image](https://user-images.githubusercontent.com/1826442/41967649-5454ce36-7a02-11e8-8d2a-953a3464b489.png)

**Comment by ujovlado** [comment]
@jancermak 

> u toho rejected ma byt jeste "Request Publishing" ?? tak jak je to ted?

ano ma to tam byt aj u rejected.

> Nebo jinak jak udelam z rejectovanyho stav request publishing? 

posles to na schvalenie a neriesime, ci si nieco zmenil.

**Comment by ujovlado** [comment]
@jancermak Komponenta je pre neho rejected pokial ju neschvalime alebo pokial ju neposle znovu na schvalenie (waiting for publishing). S @JakubMatejka sme sa dohodli, ze zatial nebudeme riesit, ktora verzia je rejectnuta alebo tak.
 
Avsak akonahle ju userposle na schvalanie, musi cakat na nas, nemoze ju poslat znovu  proste ako doteraz.

Nieco pojde podla toho ci si admin alebo trusted vendor publikovat hned, no na chovani to neriesi nic.

**Comment by jancermak** [comment]
@ujovlado @JakubMatejka 
Posilam navrh + doplneny o navrh comperovani published a last verze.

![185 - uprava logiky compare](https://user-images.githubusercontent.com/1826442/41977590-0337065e-7a20-11e8-832f-565df7dba29e.png)
![185 - uprava logiky components](https://user-images.githubusercontent.com/1826442/41977591-03548b7a-7a20-11e8-89d5-a3b66dcf7dd4.png)
![185 - uprava logiky trusted vendors](https://user-images.githubusercontent.com/1826442/41977592-037587da-7a20-11e8-81b2-b624c98bc7d4.png)

**Comment by JakubMatejka** [comment]
No, mně přijde lepší když je hned v detailu vidět ten rozdíl publishnuté vs. poslední verze. Co ty na to @ujovlado ?

**Comment by jancermak** [comment]
@ujovlado — uzivatelsky ty zmeny tam budou vyznacene, nemel bys to hledat. Budu scrolovat a uvidim 
![image](https://user-images.githubusercontent.com/1826442/42865246-5cfe51a2-8a69-11e8-9cae-71fa12d32d2e.png)

Programatorsky je to asi pracnejsi.

Ale souhlasim s tim zacit diff verzi a kdyz to bude rozumne —  jit do comparu.
